### PR TITLE
Use table ownership heuristics for migrating ACLs

### PR DESCRIPTION
**Pull request by nfx**
This PR prepends `OWN` grants when migrating ACLs for tables. For detailed logic on ownership heuristics, see #3066

### CODE REVIEW

**Comment by nfx** [comment]
fails integration tests

**Comment by nfx** [comment]
need local scheduler to improve on debuggability

```
import logging

from databricks.labs.ucx.contexts.workflow_task import RuntimeContext
from databricks.labs.ucx.runtime import Workflows

logger = logging.getLogger(__name__)


class InProcessDeployedWorkflows:
    def __init__(self, ctx: RuntimeContext):
        self._workflows = {workflow.name: workflow for workflow in Workflows.definitions()}
        self._ctx = ctx

    def run_workflow(self, step: str, **_):
        workflow = self._workflows[step]
        incoming = {task.name: 0 for task in workflow.tasks()}
        queue = []
        for task in workflow.tasks():
            task.workflow = workflow.name
            incoming[task.name] += len(task.depends_on)
        for task in workflow.tasks():
            if incoming[task.name] == 0:
                queue.append(task)
        while queue:
            task = queue.pop(0)
            fn = getattr(workflow, task.name)
            fn(self._ctx)
            for dep in task.depends_on:
                incoming[dep] -= 1
                if incoming[dep] == 0:
                    queue.append(dep)

    def relay_logs(self, step: str):
        pass  # noop

```

File: src/databricks/labs/ucx/contexts/application.py [hunk 1]
```diff
@@ -316,9 +320,15 @@ def acl_migrator(self):
             self.config.inventory_database,
         )
 
+    @cached_property
+    def table_ownership_grant_loader(self) -> TableOwnershipGrantLoader:
+        return TableOwnershipGrantLoader(self.tables_crawler, self.table_ownership)
+
     @cached_property
     def migrate_grants(self) -> MigrateGrants:
+        # owner grants have to come first
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/contexts/application.py:329]: Why?

File: src/databricks/labs/ucx/contexts/application.py [hunk 2]
```diff
@@ -316,9 +320,15 @@ def acl_migrator(self):
...
         grant_loaders: list[Callable[[], Iterable[Grant]]] = [
+            self.table_ownership_grant_loader.load,
             self.grants_crawler.snapshot,
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/contexts/application.py:332]: Should the OWN grants in this crawler be ignored?

**Review by JCZuurmond — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/contexts/application.py (hunk 1 above)
Comment (nfx) [reply src/databricks/labs/ucx/contexts/application.py:329]: because of `match_grants` returning the first matched grant

Comment (JCZuurmond) [reply src/databricks/labs/ucx/contexts/application.py:329]: Please add that to the comment

**Pull request closed without merging**